### PR TITLE
Adds possibility of scrolling to indexpaths (children included) to the HUBViewController.

### PR DESCRIPTION
**Pull request by Goos**
Putting this up as work-in-progress for discussion purposes! 

This PR adds the possibility of scrolling to nested children. 

The first challenge with doing this is that the HubFramework doesn't really know anything about how its children handles _their_ children, which forces us to delegate that to the components themselves. 

The second challenge is animations – since the components are reported to the HubFramework when they are visible, we have to wait for the animations to finish each step along the way.

@JohnSundell, @cerihughes, @spotify/objc-dev

### CODE REVIEW

**Comment by Goos** [comment]
Not entirely sure if we should keep the `scrollPosition` parameter to the HUBComponentWithChildren-call or just let them decide where to put it themselves

File: include/HubFramework/HUBComponentWithChildren.h [hunk 1]
```diff
@@ -110,6 +110,21 @@ NS_ASSUME_NONNULL_BEGIN
...
+/**
+ * Called when programmatically scrolling to a child within this parent component.
+ *
+ * @param childIndex The index of the component that is being scrolled to.
+ * @param scrollPosition The preferred position of the component after scrolling.
+ * @param animated Whether or not the scrolling should be animated.
+ * @param completionHandler The block to call once the component is visible.
+ */
+- (void)scrollToComponentAtIndex:(NSUInteger)childIndex
+                atScrollPosition:(UICollectionViewScrollPosition)scrollPosition
```
Comment (Goos) [inline include/HubFramework/HUBComponentWithChildren.h]: Not entirely sure if we should keep this parameter or just let them decide where to put it themselves

File: include/HubFramework/HUBComponentWithChildren.h [hunk 2]
```diff
@@ -110,6 +110,21 @@ NS_ASSUME_NONNULL_BEGIN
  */
 @property (nonatomic, weak, nullable) id<HUBComponentChildDelegate> childDelegate;
 
+@optional
```
Comment (JohnSundell) [inline include/HubFramework/HUBComponentWithChildren.h]: Don't use optional methods 😱 They're not compile time safe and make the API loose its explicitness. In fact, as part of https://github.com/spotify/HubFramework/issues/70 we will start linting for `@optional` and fail the build if they're used. Instead, either make it a required method (if it should be strongly coupled with this API), or make a new protocol that extends `HUBComponent`.

**Comment by Goos** [comment]
@cerihughes, @JohnSundell do you think it makes sense to call the `scrollingWillStartInViewController:` & `scrollingDidEndInViewController:` methods on the scroll handler when manually scrolling to a component indexpath?

**Comment by JohnSundell** [comment]
@Goos It probably does, because we can't make the assumption that the scrolling is both performed and observed in the same piece of code. That is, some logic in an app might cause the view to scroll, while the scroll handler for the current view is implemented in a totally different code base.

File: sources/HUBComponentWrapper.m [hunk 3]
```diff
@@ -307,6 +314,21 @@ - (void)updateViewForSelectionState:(HUBComponentSelectionState)selectionState
...
+
+- (void)scrollToComponentAtIndex:(NSUInteger)childIndex
+                  scrollPosition:(UICollectionViewScrollPosition)scrollPosition
+                        animated:(BOOL)animated
+                      completion:(void (^)())completionHandler
+{
+    if ([self.component conformsToProtocol:@protocol(HUBComponentWithScrolling)]) {
+        [(id<HUBComponentWithScrolling>)self.component scrollToComponentAtIndex:childIndex
+                                                                 scrollPosition:scrollPosition
+                                                                      animated:animated
```
Comment (ShervinS) [inline sources/HUBComponentWrapper.m]: Indentation? :D

File: sources/HUBViewControllerImplementation.m [hunk 4]
```diff
@@ -1314,6 +1337,103 @@ - (void)removeComponentWrapperFromLookupTables:(nullable HUBComponentWrapper *)c
     }
 }
 
+- (void)scrollToRootComponentAtIndex:(NSUInteger)componentIndex
+                      scrollPosition:(UICollectionViewScrollPosition)scrollPosition
+                            animated:(BOOL)animated
+                          completion:(void (^)())completionHandler
+{
+    NSUInteger const componentCount = (NSUInteger)[self.collectionView numberOfItemsInSection:0];
+    if (componentIndex == NSNotFound || componentIndex >= componentCount) {
```
Comment (Goos) [inline sources/HUBViewControllerImplementation.m]: @JohnSundell – do we normally assert these things or fail silently?

Comment (JohnSundell) [inline sources/HUBViewControllerImplementation.m]: Normally, we assert - since it's an incorrect use of the API. Failing silently has the downside that no-one knows that they did something wrong, until we get a bug report 😄

**Comment by Goos** [comment]
@JohnSundell, @cerihughes unit tests are in place – would you mind taking another look? :smile:

File: include/HubFramework/HUBComponentWithScrolling.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Called when programmatically scrolling to a child within this parent component.
+ *
+ * @param childIndex The index of the component that is being scrolled to.
+ * @param scrollPosition The preferred position of the component after scrolling.
+ * @param animated Whether or not the scrolling should be animated.
+ * @param completionHandler The block to call once the component is visible.
+ */
+- (void)scrollToComponentAtIndex:(NSUInteger)childIndex
+                  scrollPosition:(UICollectionViewScrollPosition)scrollPosition
```
Comment (JohnSundell) [inline include/HubFramework/HUBComponentWithScrolling.h]: With this, we're kind of assuming that the component will be implemented using `UICollectionView`, which is not always true. Do we need to support multiple scroll positions at this point? If so - I'd say make a custom `enum` that is not tied to `UICollectionView`.

File: include/HubFramework/HUBComponentWithScrolling.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                        animated:(BOOL)animated
+                      completion:(void (^)())completionHandler;
```
Comment (JohnSundell) [inline include/HubFramework/HUBComponentWithScrolling.h]: Make the external parameter name match the documentation and the internal name - `completionHandler`.

File: include/HubFramework/HUBViewController.h [hunk 7]
```diff
@@ -171,6 +171,19 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)scrollToContentOffset:(CGPoint)contentOffset animated:(BOOL)animated;
 
 /**
+ *  Scroll to a desired content offset.
```
Comment (JohnSundell) [inline include/HubFramework/HUBViewController.h]: Not content offset, right? To a component at a given index path?

File: include/HubFramework/HUBViewController.h [hunk 8]
```diff
@@ -171,6 +171,19 @@ NS_ASSUME_NONNULL_BEGIN
...
+ *
+ *  @param componentIndexPath The index path of the component to scroll to.
+ *  @param scrollPosition The preferred position of the component after scrolling.
+ *  @param animated Whether or not the scrolling should be animated.
+ *  @param completionHandler A block that is called once the component at the provided index path is visible.
+ */
+- (void)scrollToComponentAtIndexPath:(NSIndexPath *)componentIndexPath
+                      scrollPosition:(UICollectionViewScrollPosition)scrollPosition
+                            animated:(BOOL)animated
+                          completion:(void (^ _Nullable)())completionHandler;
```
Comment (JohnSundell) [inline include/HubFramework/HUBViewController.h]: Same here - `completionHandler`

File: include/HubFramework/HUBViewControllerScrollHandler.h [hunk 9]
```diff
@@ -110,4 +110,15 @@
...
                                        proposedContentOffset:(CGPoint)proposedContentOffset;
 
+/**
+ * Return the content offset for displaying a component at a certain scroll position.
+ * 
+ * @param componentIndex The index of the component to display.
+ * @param scrollPosition The position to display the component at.
+ * @param viewController The view controller in question.
+ */
+- (CGPoint)contentOffsetForDisplayingComponentAtIndex:(NSUInteger)componentIndex
```
Comment (JohnSundell) [inline include/HubFramework/HUBViewControllerScrollHandler.h]: Hmmm, why is this needed? 🤔 Should the scroll handler have an opinion on where the children are?

File: sources/HUBComponentWrapper.h [hunk 10]
```diff
@@ -196,6 +198,12 @@ willUpdateSelectionState:(HUBComponentSelectionState)selectionState;
  */
 - (void)saveComponentUIState;
 
+/** 
+ *  Returns the child component wrapper located at the provided index – if visible. 
+ *  @param index The index of the component to retrieve.
```
Comment (JohnSundell) [inline sources/HUBComponentWrapper.h]: 🎨 New line between method description and `@param`

File: sources/HUBComponentWrapper.h [hunk 11]
```diff
@@ -196,6 +198,12 @@ willUpdateSelectionState:(HUBComponentSelectionState)selectionState;
...
+ */
+- (nullable HUBComponentWrapper *)childComponentAtIndex:(NSUInteger)index;
```
Comment (JohnSundell) [inline sources/HUBComponentWrapper.h]: Since this is only returning visible components - how about naming it `visibleChildComponentAtIndex:`?

File: sources/HUBComponentWrapper.m [hunk 12]
```diff
@@ -118,6 +120,11 @@ - (void)saveComponentUIState
     }
 }
 
+- (nullable HUBComponentWrapper *)childComponentAtIndex:(NSUInteger)index
+{
+    return self.childrenByIndex[@(index)];
```
Comment (JohnSundell) [inline sources/HUBComponentWrapper.m]: This doesn't only contain visible children. It contains all children created. So the docs doesn't match 🙂

File: sources/HUBComponentWrapper.m [hunk 13]
```diff
@@ -307,6 +314,21 @@ - (void)updateViewForSelectionState:(HUBComponentSelectionState)selectionState
...
 }
 
+#pragma mark - HUBComponentWithScrolling
+
+- (void)scrollToComponentAtIndex:(NSUInteger)childIndex
+                  scrollPosition:(UICollectionViewScrollPosition)scrollPosition
+                        animated:(BOOL)animated
+                      completion:(void (^)())completionHandler
+{
+    if ([self.component conformsToProtocol:@protocol(HUBComponentWithScrolling)]) {
```
Comment (JohnSundell) [inline sources/HUBComponentWrapper.m]: 🎨 Return early instead of indenting method?

File: sources/HUBViewControllerDefaultScrollHandler.m [hunk 14]
```diff
@@ -68,6 +69,26 @@ - (CGPoint)targetContentOffsetForEndedScrollInViewController:(UIViewController<H
     return proposedContentOffset;
 }
 
+- (CGPoint)contentOffsetForDisplayingComponentAtIndex:(NSUInteger)componentIndex
+                                       scrollPosition:(UICollectionViewScrollPosition)scrollPosition
+                                       viewController:(UIViewController<HUBViewController> *)viewController
+{
+    CGRect const componentFrame = [viewController frameForBodyComponentAtIndex:componentIndex];
```
Comment (JohnSundell) [inline sources/HUBViewControllerDefaultScrollHandler.m]: Really cuirous as to why this should be the responsibility of the scroll handler. Seems like something the view controller should do.

File: sources/HUBViewControllerImplementation.m [hunk 15]
```diff
@@ -1314,6 +1337,105 @@ - (void)removeComponentWrapperFromLookupTables:(nullable HUBComponentWrapper *)c
...
 }
 
+- (void)scrollToRootComponentAtIndex:(NSUInteger)componentIndex
+                      scrollPosition:(UICollectionViewScrollPosition)scrollPosition
+                            animated:(BOOL)animated
+                          completion:(void (^)())completionHandler
+{
+    NSUInteger const componentCount = (NSUInteger)[self.collectionView numberOfItemsInSection:0];
+
+    NSParameterAssert(componentIndex != NSNotFound);
```
Comment (JohnSundell) [inline sources/HUBViewControllerImplementation.m]: This is not needed as your next check will fail if the value is `NSNotFound` (since it's `INT_MAX`)

File: sources/HUBViewControllerImplementation.m [hunk 16]
```diff
@@ -1314,6 +1337,105 @@ - (void)removeComponentWrapperFromLookupTables:(nullable HUBComponentWrapper *)c
...
+    if (indexPosition > 0) {
+        NSParameterAssert(childIndex <= componentWrapper.model.children.count);
+    }
+
+    __weak HUBViewControllerImplementation *weakSelf = self;
+    void (^stepCompletionHandler)() = ^{
+        HUBViewControllerImplementation *strongSelf = weakSelf;
+        // If the scrolling isn't animated we have to wait for a runloop cycle, as scrolling is deferred.
+        if (!animated) {
+            [[NSRunLoop currentRunLoop] runUntilDate:[NSDate date]];
```
Comment (JohnSundell) [inline sources/HUBViewControllerImplementation.m]: Won't this block the main thread for the remainder of the cycle? Wouldn't it be better to `dispatch_async()`?

File: include/HubFramework/HUBViewControllerScrollHandler.h (hunk 9 above)
Comment (Goos) [inline include/HubFramework/HUBViewControllerScrollHandler.h]: I figured it made sense that the scroll handler had a say in all kinds of scrolling (i.e. programmatic and interactive scrolling) :smile:

File: include/HubFramework/HUBComponentWithScrolling.h (hunk 6 above)
Comment (tammofreese) [inline include/HubFramework/HUBComponentWithScrolling.h]: How about making the block type more concrete by using `void (^)(void)` instead of `void (^)()`?

**Comment by Goos** [comment]
@JohnSundell fixed a bunch of stuff – mind taking another look? :smile:

File: include/HubFramework/HUBViewController.h [hunk 17]
```diff
@@ -171,6 +172,28 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)scrollToContentOffset:(CGPoint)contentOffset animated:(BOOL)animated;
 
 /**
+ *  Scroll to a desired component. Each index in the index path refers to one level of children.
+ *  For example, in order to scroll to a root component at an index, you would provide an index
+ *  path with that single index. If that component in turn has children, you can scroll between
+ *  those by providing an index path with two index, starting with the root index, and so on.
+ *
+ *  In order for child components to support nested scrolling, they must implement @c HUBComponentWithScrolling.
```
Comment (krris) [inline include/HubFramework/HUBViewController.h]: What is `@c`?

File: include/HubFramework/HUBComponentWithScrolling.h [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+#import "HUBComponentWithChildren.h"
+#import "HUBScrollPosition.h"
+
+/**
+ * Extended Hub component protocol that adds the ability scroll between child components.
```
Comment (krris) [inline include/HubFramework/HUBComponentWithScrolling.h]: 🔬  ... ability **to** scroll

File: include/HubFramework/HUBScrollPosition.h [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+#import <UIKit/UIKit.h>
+
+/**
+ * An enumeration to be used when scrolling to a specific position within a scrollable component.
+ * For ease of use, the different options map to their corresponding UICollectionViewScrollPosition.
```
Comment (JohnSundell) [inline include/HubFramework/HUBScrollPosition.h]: Smart! 👍

File: include/HubFramework/HUBViewController.h [hunk 20]
```diff
@@ -171,6 +172,28 @@ NS_ASSUME_NONNULL_BEGIN
...
+ *
+ *  @param componentType The type of component you want to scroll to.
+ *  @param indexPath The index path of the component to scroll to.
+ *  @param scrollPosition The preferred position of the component after scrolling.
+ *  @param animated Whether or not the scrolling should be animated.
+ *  @param completion A block that is called once the component at the provided index path is visible.
+ *
+ *  @seealso HUBComponentWithScrolling
+ */
+- (void)scrollToComponentOfType:(HUBComponentType)componentType
```
Comment (JohnSundell) [inline include/HubFramework/HUBViewController.h]: Awesome 🚀

File: include/HubFramework/HUBViewControllerScrollHandler.h (hunk 9 above)
Comment (JohnSundell) [inline include/HubFramework/HUBViewControllerScrollHandler.h]: If we don't need this API - why add it? 😄 It just makes the scroll handler more complicated. Why can't we use the normal content offset calculating method? 🙂

File: include/HubFramework/HUBViewController.h (hunk 17 above)
Comment (Goos) [inline include/HubFramework/HUBViewController.h]: It gives the following word a monospaced font :smile:
